osquery/tables/system/windows/drivers.cpp

```diff
@@ -0,0 +1,199 @@
+#include <Winsvc.h>
+#include <psapi.h>
+#include <string>
+
+#include <osquery/core.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/tables/system/windows/registry.h"
+
+#pragma comment(lib, "Advapi32.lib")
+
+namespace osquery {
+namespace tables {
+
+const std::string kDrvStartType[] = {
+    "BOOT_START", "SYSTEM_START", "AUTO_START", "DEMAND_START", "DISABLED"};
+
+const std::string kDrvStatus[] = {"UNKNOWN",
+                                  "STOPPED",
+                                  "START_PENDING",
+                                  "STOP_PENDING",
+                                  "RUNNING",
+                                  "CONTINUE_PENDING",
+                                  "PAUSE_PENDING",
+                                  "PAUSED"};
+
+const std::map<int, std::string> kDriverType = {
+    {0x00000001, "KERNEL"}, {0x00000002, "FILE_SYSTEM"},
+};
+
+static std::map<std::string, std::string> loadedDrivers;
+
+bool queryDrvInfo(const SC_HANDLE& schScManager,
+                  ENUM_SERVICE_STATUS_PROCESS& svc,
+                  Row& r) {
+  DWORD cbBufSize = 0;
+
+  auto schService =
+      OpenService(schScManager, svc.lpServiceName, SERVICE_QUERY_CONFIG);
+
+  if (schService == nullptr) {
+    TLOG << "OpenService failed (" << GetLastError() << ")";
+    return false;
+  }
+
+  QueryServiceConfig(schService, nullptr, 0, &cbBufSize);
+  auto lpsc = (LPQUERY_SERVICE_CONFIG)malloc(cbBufSize);
+  if (QueryServiceConfig(schService, lpsc, cbBufSize, &cbBufSize) != 0) {
+    TLOG << "QueryServiceConfig failed (" << GetLastError() << ")";
+  }
+
+  // SCM can provide more info about the driver but not all drivers are
+  // managed by SCM, So Here we remove driver from loadedDrivers list to avoid
+  // duplicates
+  // As the driver is already in SCM list.
+  loadedDrivers.erase(lpsc->lpBinaryPathName);
+
+  r["name"] = SQL_TEXT(svc.lpServiceName);
+  r["display_name"] = SQL_TEXT(svc.lpDisplayName);
+  r["status"] = SQL_TEXT(kDrvStatus[svc.ServiceStatusProcess.dwCurrentState]);
+  r["start_type"] = SQL_TEXT(kDrvStartType[lpsc->dwStartType]);
+  r["path"] = SQL_TEXT(lpsc->lpBinaryPathName);
+
+  if (kDriverType.count(lpsc->dwServiceType) > 0) {
+    r["driver_type"] = SQL_TEXT(kDriverType.at(lpsc->dwServiceType));
+  } else {
+    r["driver_type"] = SQL_TEXT("UNKNOWN");
+  }
+
+  QueryData regResults;
+  queryKey("HKEY_LOCAL_MACHINE",
+           "SYSTEM\\CurrentControlSet\\Services\\" + r["name"],
+           regResults);
+  for (const auto& aKey : regResults) {
+    if (aKey.at("name") == "Owners") {
+      r["inf"] = SQL_TEXT(aKey.at("data"));
+    }
+  }
+
+  free(lpsc);
+  CloseServiceHandle(schService);
+  return true;
+}
+
+void enumLoadedDrivers() {
+  DWORD bytesNeeded = 0;
+  int driversCount = 0;
+
+  auto ret = EnumDeviceDrivers(nullptr, 0, &bytesNeeded);
+  auto drvBaseAddr = (LPVOID*)malloc(bytesNeeded);
+
+  ret = EnumDeviceDrivers(drvBaseAddr, bytesNeeded, &bytesNeeded);
```

Comment:
@theopolis do you mean checking if drvBaseAddr is not null ?

i did a small test if `drvBaseAddr` is null
if debugger attached i got an exception when calling `EnumDeviceDrivers`
if i just run from cmd i got everything running without exceptions.
So, i will add check for NULL pointer to avoid any kinds of exceptions